### PR TITLE
Allow custom validation response

### PR DESCRIPTION
**Pull request by hhamana**
This PR fixes 2 issues:
- A user sending invalid data in a UUID-based plugin should not result in a server error, but a client error with a 400 status code.
- Middlewares in Starlette should not raise Exceptions, they should abort the request cycle and send a response instead.

The exact error is customizable from user code, but provides a default 400 response with empty body.
Implemented on both ContextMiddleware and RawContextMiddleware.

### CODE REVIEW

**Comment by tomwojcik** [comment]
Thank you for your contribution! Within the next few days I'm afraid I won't have time to look into this PR but I will try to find some time later on.

File: starlette_context/middleware/context_middleware.py [hunk 1]
```diff
@@ -22,12 +22,18 @@ class ContextMiddleware(BaseHTTPMiddleware):
     """
 
     def __init__(
-        self, plugins: Optional[Sequence[Plugin]] = None, *args, **kwargs
+        self,
+        plugins: Optional[Sequence[Plugin]] = None,
+        defaut_error_response: Response = Response(status_code=400),
```
Comment (tomwojcik) [inline starlette_context/middleware/context_middleware.py:27]: How is this problem solved in other middlewares?

I think this parameter should be a callback function instead. Maybe accept `Exception` and `Request` and based on these two compose error message? From the other hand I don't think it's a good idea at all because there already is exception handler that does exactly that, but not for 5xx. 🤔

File: starlette_context/middleware/context_middleware.py [hunk 2]
```diff
@@ -43,15 +49,22 @@ async def set_context(self, request: Request) -> dict:
     async def dispatch(
         self, request: Request, call_next: RequestResponseEndpoint
     ) -> Response:
-        _starlette_context_token: Token = _request_scope_context_storage.set(
-            await self.set_context(request)
-        )
+        try:
+            context = await self.set_context(request)
+            token: Token = _request_scope_context_storage.set(context)
+        except ValueError as e:
```
Comment (tomwojcik) [inline starlette_context/middleware/context_middleware.py:55]: I'm still not sure if it's a good idea to handle it like that but if we were to proceed it can't be `ValueError` as something else might raise it. Only errors we know we want to handle here should be handled, otherwise 500 is correct. On master you can find `errors.py`. Maybe it's time to introduce `StarletteContextError` base class and the raise per plugin error such as `WrongUUIDError(StarletteContextError)`?

File: starlette_context/middleware/context_middleware.py [hunk 3]
```diff
@@ -43,15 +49,22 @@ async def set_context(self, request: Request) -> dict:
...
+            if isinstance(e.args[0], Response):
+                error_response = e.args[0]
```
Comment (tomwojcik) [inline starlette_context/middleware/context_middleware.py:57]: That's hacky. As I said, I'd prefer if the error message was created in a single place, in the class, something like `handle_plugin_error(self, exc, request)`. wdyt?

File: starlette_context/middleware/raw_middleware.py [hunk 4]
```diff
@@ -54,11 +61,31 @@ async def send_wrapper(message: Message) -> None:
...
+                error_response = e.args[0]
+            else:
+                error_response = self.error_response
+
+            message_head = {
+                "type": "http.response.start",
+                "status": error_response.status_code,
+                "headers": error_response.raw_headers,
+            }
+            await send(message_head)
```
Comment (tomwojcik) [inline starlette_context/middleware/raw_middleware.py:77]: I know what it's supposed to do but I can't get use to this async weirdness. Can you link to another middleware that does a similar thing?

File: starlette_context/plugins/base.py [hunk 5]
```diff
@@ -52,18 +52,20 @@ def __init__(
         force_new_uuid: bool = False,
         version: int = 4,
         validate: bool = True,
+        error_response: Optional[Response] = None,
```
Comment (tomwojcik) [inline starlette_context/plugins/base.py:55]: I think this callback func mentioned above could create a response based on the error, wdyt?

File: starlette_context/plugins/base.py [hunk 6]
```diff
@@ -52,18 +52,20 @@ def __init__(
...
         self.version = version
         self.validate = validate
+        self.error_response = error_response
 
     def validate_uuid(self, uuid_to_validate: str) -> None:
         try:
             uuid.UUID(uuid_to_validate, version=self.version)
         except Exception as e:
-            raise ValueError("Wrong uuid") from e
+            raise ValueError(self.error_response or "Wrong uuid") from e
```
Comment (tomwojcik) [inline starlette_context/plugins/base.py:68]: I think it should just raise `WrongUUIDError(StarletteContextError)` from `errors.py`.

File: tests/test_plugins/test_error_responses.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    ]
+    app = Starlette(middleware=middleware)
+    client = TestClient(app)
+    return client
+
+
+def test_invalid_request_id_returns_specified_response_raw_middleware():
+    content = {"Error": "Invalid X-Request-ID"}
+    response = JSONResponse(
+        content=content, status_code=status.HTTP_422_UNPROCESSABLE_ENTITY
```
Comment (tomwojcik) [inline tests/test_plugins/test_error_responses.py:34]: I am also thinking it should be either 422 or 400. I'm not sure as it's a header, so kind of a request metadata. What's your reasoning?

**Review by tomwojcik — CHANGES_REQUESTED** [review]

**Comment by tomwojcik** [comment]
Sorry it took so long, I have a few personal things going on.

So this PR looks really well. You're right it'd be better if 4xx was returned instead.
It seems like you're right with the exceptions from the middleware as well.

I have a few comments though. Do you want me to accept it as is and change a few things or do you want to discuss comments?

File: starlette_context/middleware/context_middleware.py (hunk 3 above)
Comment (hhamana) [reply starlette_context/middleware/context_middleware.py:57]: yes, this part is quite hacky indeed. My original attempt was to define the error response on the plugin so the error message could be fine-grained, before moving back up to have the default_error_response at the middleware level.
Fundamentally, the options for the plugin to communicate back to the middleware are quite limited, so I opted to put the response itself as Exception argument.
It would make sense to define it from the middleware exclusively, but that would tend to enforce a generic catch-all error response. This allows both fine-grained and generic approaches.
Passing a callback function might also be a possibility, but it makes it slightly more annoying to implement on user side, and might feel like a small re-implementation of the error handling system.

File: starlette_context/middleware/context_middleware.py (hunk 2 above)
Comment (hhamana) [reply starlette_context/middleware/context_middleware.py:55]: I agree, it's better to have a specific Exception type, and it might be useful if people raise errors from custom plugin subclasses we don't control.

File: starlette_context/plugins/base.py (hunk 5 above)
Comment (hhamana) [reply starlette_context/plugins/base.py:55]: It seems this is the approach taken by the the ASGI RateLimit middleware 
https://github.com/abersheeran/asgi-ratelimit/blob/master/ratelimit/core.py
taking the callback from init (I think that makes it a signature similar to `def callback(exc: Exception) -> Response`):
```
        on_auth_error: Optional[Callable[[Exception], Awaitable[ASGIApp]]] = None,
```
and using it on error cases
```
                    except Exception as exc:
                        if self.on_auth_error is not None:
                            reponse = await self.on_auth_error(exc)
                            return await reponse(scope, receive, send)
                        raise exc
```
I guess that might work, although completely shifts the burden on users to define the callback and only gives the exception.
With a custom exception, that might be acceptable...

File: starlette_context/plugins/base.py (hunk 6 above)
Comment (hhamana) [reply starlette_context/plugins/base.py:68]: Agreed, at least it would be applying a best practice, at best makes it easier to separate issues that might come from user-defined plugin subclasses

File: tests/test_plugins/test_error_responses.py (hunk 7 above)
Comment (hhamana) [reply tests/test_plugins/test_error_responses.py:34]: default is a 400, which I think is a fair default, and this tests that user can change it to whatever they want if they disagree.

**Comment by hhamana** [comment]
Thank you for the review. I agree there should be some changes as per your comments, I'll experiment with a callback-based implementation.

File: tests/test_plugins/test_error_responses.py (hunk 7 above)
Comment (hhamana) [reply tests/test_plugins/test_error_responses.py:34]: as for the reasoning, I would personally prefer to assign 422 to body validation as they would represent objects, which would be close enough to the 422 "entity". 400 is the generic client-side so this fits better as a header-related validation default imo. But this is obviously something that people should be left free to choose.

**Comment by tomwojcik** [comment]
> I don't think it's very user-friendly to make users implement that sort of thing, and would prefer the response be generated here as it is, with the bonus of offering a consistent interface for the different middlewares.

Ehh... I was thinking about one thing and wrote something else. Of course, it's not user-friendly. A huge part of my review was not entirely correct, especially the callback/base exception thing. Now I gave it a thought and I think there's only one correct way of handling those.

To my knowledge, all web frameworks have a central place where they handle errors. Starlette is a framework so it imposes a certain way of handling them as well, in the [ExceptionMiddleware](https://github.com/encode/starlette/blob/4750e7b6135e7760560938aa6c9699dd48ff2887/starlette/applications.py#L79-L87). 

If you had middleware A, B , C and RawContextMiddleware, create a response in RawContextMiddleware, the rest of them will process the response as if it was a success response. So for instance ASGISentryMiddleware etc. wouldn't catch the error. It's just asking for trouble. We need to do things Starlette-way.

1. `StarletteContextError` is not needed. `Starlette` is responsible for handling errors and `starlette-context` has to raise errors `Starlette` can understand. All custom errors should inherit from [`HTTPException`](https://github.com/encode/starlette/blob/4750e7b6135e7760560938aa6c9699dd48ff2887/starlette/exceptions.py#L11) and have default arguments. 
Example:
```python
class InvalidUUIDError(HTTPException): 
    def __init__(self, status_code: int = status.HTTP_422_UNPROCESSABLE_ENTITY, detail: str = 'Invalid UUID') -> None: 
        super().__init__(status_code, detail)
```
2. Now there's no need for passing error body such as this one
```python
          raise WrongUUIDError(  # should be default
                "Wrong uuid", error_response=self.error_response
            ) from e
```
3. ExceptionMiddleware [handles](https://github.com/encode/starlette/blob/4750e7b6135e7760560938aa6c9699dd48ff2887/starlette/exceptions.py#L90) HTTPException, which basically means it creates the `Response`.

 Because of that:
 - there's no need to accept response argument in plugins and middleware as they don't create any
 - there's no need to handle ANY exception in middlewares. Non-HTTPException should result in 5xx.
 - there's no need to do this `await send` weirdness as Starlette will handle that as well
 - `starlette-context` doesn't need to learn how it works. They just need to learn how exceptions in `Starlette` are handled (if they want to change a few things)

To recap, the problem you noticed is that Python built-in errors are raised which results in 5xx. As those are request-response errors, they can be handled by the framework. The only thing that has to be done is changing existing errors to `HTTPException` errors (or adding new ones) so the framework knows how to create a relevant response for the client.

What do you think?

**Comment by hhamana** [comment]
I had a slight concern that raising as an HTTPException would make it hard to customize the response on user side, but I think most apps would define their own error handling anyway, and this would integrate it nicely along with Starlette and such errors. That approach should be sufficient, and indeed simpler here too.

**Comment by hhamana** [comment]
My attempts to implement this based on HTTPException failed, for the exact same reason for the original implementation.
As per official Starlette documentation:
> You should only raise HTTPException inside routing or endpoints. Middleware classes should instead just return appropriate responses directly.

The response for this is, again from Starlette documentation:
>  the middleware stack of a Starlette application is configured like this:
    - ServerErrorMiddleware - Returns 500 responses when server errors occur.
    - Installed middleware
    - ExceptionMiddleware - Deals with handled exceptions, and returns responses
    - Router
    - Endpoints
Due to this design, anything that is raised from a middleware will only bubble up to the `ServerErrorMiddleware`, which is only capable of 500 responses (or dirty hacks to override this and handle incorrect cases).
Middlewares MUST NOT raise. The HTTPException is no exception, as is it handled in the `ExceptionMiddleware` below in the stack.

The current implementation is correct.

**Comment by tomwojcik** [comment]
@hhamana I remember about your PR but I'm currently overwhelmed with non-side-projects responsibilities. I'm grateful for your contribution but I can't look into it right now. Sorry for that.

**Comment by tomwojcik** [comment]
I still don't like this optional response here and there. I was playing with your branch and indeed I think what you did is the only way to handle plugin (middleware) error.

This PR looks really good. Thank you for your contribution and active discussion (and patience :D).

**Comment by hhamana** [comment]
Thank you very much.
I totally understand your feelings, sending a response object through the exception is weird, but I can't see any other way to make it work nicely with plugins.
Now that we agreed on that, I think some documentation would be in order. I hope I can write something soon-ish.
